src/jit/optimizer.cpp

```diff
@@ -6795,8 +6795,18 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
 
         bool IsTreeVNInvariant(GenTree* tree)
         {
-            return m_compiler->optVNIsLoopInvariant(tree->gtVNPair.GetLiberal(), m_loopNum,
-                                                    &m_hoistContext->m_curLoopVnInvariantCache);
+            ValueNum vn = tree->gtVNPair.GetLiberal();
+
+            if (m_compiler->vnStore->IsVNConstant(vn))
+            {
+                // Don't allow the hoisting of GT_CLS_VAR an GT_LCL_VAR that were assigned a constant value.
```

Comment:
This comment is not accurate: we are not allowing any nodes (not just GT_CLS_VAR and GT_LCL_VAR) except GT_ARGPLACE.

Reply:
I will try to clarify the comment.  But yes, the intention is to restrict the nodes that can be short-circuit later to just be plain constant nodes or the GT_ARGPLACE nodes.

Reply:
Updated to:

```
            if (m_compiler->vnStore->IsVNConstant(vn))
            {
                // We only allow true constant nodes  (i.e. GT_CNS_INT, etc...) or a GT_ARGPLACE nodes.
                // It is unsafe to allow a GT_CLS_VAR that has been assigned a constant.
                // The logic in optVNIsLoopInvariant woudl consider it to be loop-invariant, even
                // if the assignment of the constant to the GT_CLS_VAR was inside the loop.
                //
                if (!tree->OperIsConst() && tree->OperGet() != GT_ARGPLACE)
                {
                    return false;
                }
            }
```